optimum/habana/transformers/generation/utils.py

```diff
@@ -189,6 +192,116 @@ def _update_model_kwargs_for_generation(
 
         return model_kwargs
 
+    def _prepare_decoder_attention_mask(
+        self,
+        max_steps: int,  # current stopping criteria
+        batch_size: int,
+        pad_token_id: int,
+        device: str,
+        dtype: str = torch.int64,
+    ) -> torch.Tensor:
+        x = torch.zeros((batch_size, max_steps), device=device, dtype=dtype)
+        return x.index_fill(1, torch.tensor([0]), 1)  # First the position with pad_token_id
+
+    def _prepare_past_key_values(
+        self,
+        max_steps: int,  # current stopping criteria
+        batch_size: int,
+        device: str,
+    ) -> Tuple[torch.LongTensor, torch.LongTensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        n_heads = self.config.decoder_attention_heads
+        embed_size_per_head = self.config.d_model // n_heads
+        key_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        value_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        past_key_values = tuple(
+            (key_states_[i, :, :, :, :], value_states_[i, :, :, :, :], None, None)
+            for i in range(self.config.decoder_layers)
+        )
+        return past_key_values
+
+    def _prepare_decoder_input_ids_for_generation(
+        self,
+        batch_size: int,
+        model_input_name: str,
+        model_kwargs: Dict[str, torch.Tensor],
+        decoder_start_token_id: int = None,
+        bos_token_id: int = None,
+        device: torch.device = None,
+        max_length: int = 1,
+    ) -> Tuple[torch.LongTensor, Dict[str, torch.Tensor]]:
```

Comment:
Since this method overrides the one of Transformers, could you write a short docstring that tells the difference with the Transformers one? It can be just one sentence. It seems to be linked to the initialization of `decoder_input_ids_start` if I'm not mistaken.